File: SkipList/SkipList.ts
/**
 * SkipList implementation with probabilistic level assignment.
 * Provides O(log n) average time for search, insert, and delete operations.
 */
class SkipListNode {
    key: number;
    forward: (SkipListNode | null)[];

    constructor(key: number, level: number) {
        this.key = key;
        this.forward = new Array(level + 1).fill(null);
    }
}

class SkipList {
    private maxLevel: number;
    private p: number;
    private header: SkipListNode;
    private level: number;

    constructor(maxLevel: number = 16, p: number = 0.5) {
        this.maxLevel = maxLevel;
        this.p = p;
        this.header = new SkipListNode(-Infinity, maxLevel);
        this.level = 0;
    }

    /**
     * Generate random level for new node.
     */
    private randomLevel(): number {
        let level = 0;
        while (Math.random() < this.p && level < this.maxLevel) {
            level++;
        }
        return level;
    }

    /**
     * Search for key in SkipList. O(log n) average time.
     */
    search(key: number): number | null {
        let current = this.header;
        for (let i = this.level; i >= 0; i--) {
            while (current.forward[i] && current.forward[i]!.key < key) {
                current = current.forward[i]!;
            }
        }

        current = current.forward[0]!;
        if (current && current.key === key) {
            return current.key;
        }
        return null;
    }

    /**
     * Insert key into SkipList. O(log n) average time.
     */
    insert(key: number): void {
        const update: (SkipListNode | null)[] = new Array(this.maxLevel + 1).fill(null);
        let current = this.header;

        for (let i = this.level; i >= 0; i--) {
            while (current.forward[i] && current.forward[i]!.key < key) {
                current = current.forward[i]!;
            }
            update[i] = current;
        }

        current = current.forward[0]!;

        if (!current || current.key !== key) {
            const newLevel = this.randomLevel();

            if (newLevel > this.level) {
                for (let i = this.level + 1; i <= newLevel; i++) {
                    update[i] = this.header;
                }
                this.level = newLevel;
            }

            const newNode = new SkipListNode(key, newLevel);

            for (let i = 0; i <= newLevel; i++) {
                newNode.forward[i] = update[i]!.forward[i];
                update[i]!.forward[i] = newNode;
            }
        }
    }

    /**
     * Delete key from SkipList. O(log n) average time.
     */
    delete(key: number): void {
        const update: (SkipListNode | null)[] = new Array(this.maxLevel + 1).fill(null);
        let current = this.header;

        for (let i = this.level; i >= 0; i--) {
            while (current.forward[i] && current.forward[i]!.key < key) {
                current = current.forward[i]!;
            }
            update[i] = current;
        }

        current = current.forward[0]!;

        if (current && current.key === key) {
            for (let i = 0; i <= this.level; i++) {
                if (update[i]!.forward[i] !== current) {
                    break;
                }
                update[i]!.forward[i] = current.forward[i];
            }

            while (this.level > 0 && !this.header.forward[this.level]) {
                this.level--;
            }
        }
    }

    /**
     * Display SkipList structure.
     */
    display(): void {
        console.log("SkipList structure:");
        for (let level = this.level; level >= 0; level--) {
            let current = this.header.forward[level];
            let levelStr = `Level ${level}: `;
            while (current) {
                levelStr += `${current.key} -> `;
                current = current.forward[level];
            }
            levelStr += "None";
            console.log(levelStr);
        }
    }
}

/**
 * Demonstrates SkipList operations with probabilistic balancing.
 * Inserts 10 elements, searches for values, and displays structure.
 */
function main(): void {
    const startTime = performance.now();
    const memoryBefore = process.memoryUsage().heapUsed;
    
    // Create SkipList instance
    const skipList = new SkipList();
    
    // Insert elements
    const elements = [3, 6, 7, 9, 12, 19, 17, 26, 21, 25];
    for (const elem of elements) {
        skipList.insert(elem);
    }

    // Display structure
    skipList.display();
    
    // Search operations
    const searchKeys = [19, 15, 21];
    console.log("\nSearch results:");
    for (const key of searchKeys) {
        const result = skipList.search(key);
        console.log(`Search for ${key}: ${result ? 'Found' : 'Not found'}`);
    }
    
    // Delete operation
    skipList.delete(17);
    console.log("\nAfter deleting 17:");
    skipList.display();
    
    const endTime = performance.now();
    const memoryAfter = process.memoryUsage().heapUsed;
    const memoryUsed = memoryAfter - memoryBefore;
    
    // Performance statistics
    console.log("\n--- Performance Statistics ---");
    console.log(`Execution time: ${(endTime - startTime).toFixed(4)} ms`);
    console.log(`Memory usage: ${memoryUsed > 0 ? memoryUsed : 'negligible'} bytes`);
    console.log("\nTime Complexity (Average):");
    console.log("  - Search: O(log n)");
    console.log("  - Insert: O(log n)");
    console.log("  - Delete: O(log n)");
    console.log("  - Space: O(n)");
}

main();

File: SkipList/SkipList.java
package SkipList;

import java.util.Random;

/**
 * SkipList implementation with probabilistic level assignment.
 * Provides O(log n) average time for search, insert, and delete operations.
 * Demonstrates SkipList operations with probabilistic balancing.
 * Inserts 10 elements, searches for values, and displays structure.
 */
public class SkipList {

    private static class SkipListNode {
        int key;
        SkipListNode[] forward;

        SkipListNode(int key, int level) {
            this.key = key;
            this.forward = new SkipListNode[level + 1];
        }
    }

    private static class SkipListImpl {
        private static final int MAX_LEVEL = 16;
        private static final double P = 0.5;
        private SkipListNode header;
        private int level;
        private Random random;

        public SkipListImpl() {
            this.header = new SkipListNode(Integer.MIN_VALUE, MAX_LEVEL);
            this.level = 0;
            this.random = new Random();
        }

        private int randomLevel() {
            int level = 0;
            while (random.nextDouble() < P && level < MAX_LEVEL) {
                level++;
            }
            return level;
        }

        /**
         * Search for key in SkipList. O(log n) average time.
         */
        public Integer search(int key) {
            SkipListNode current = header;
            for (int i = level; i >= 0; i--) {
                while (current.forward[i] != null && current.forward[i].key < key) {
                    current = current.forward[i];
                }
            }

            current = current.forward[0];
            if (current != null && current.key == key) {
                return current.key;
            }
            return null;
        }

        /**
         * Insert key into SkipList. O(log n) average time.
         */
        public void insert(int key) {
            SkipListNode[] update = new SkipListNode[MAX_LEVEL + 1];
            SkipListNode current = header;

            for (int i = level; i >= 0; i--) {
                while (current.forward[i] != null && current.forward[i].key < key) {
                    current = current.forward[i];
                }
                update[i] = current;
            }

            current = current.forward[0];

            if (current == null || current.key != key) {
                int newLevel = randomLevel();

                if (newLevel > level) {
                    for (int i = level + 1; i <= newLevel; i++) {
                        update[i] = header;
                    }
                    level = newLevel;
                }

                SkipListNode newNode = new SkipListNode(key, newLevel);

                for (int i = 0; i <= newLevel; i++) {
                    newNode.forward[i] = update[i].forward[i];
                    update[i].forward[i] = newNode;
                }
            }
        }

        /**
         * Delete key from SkipList. O(log n) average time.
         */
        public void delete(int key) {
            SkipListNode[] update = new SkipListNode[MAX_LEVEL + 1];
            SkipListNode current = header;

            for (int i = level; i >= 0; i--) {
                while (current.forward[i] != null && current.forward[i].key < key) {
                    current = current.forward[i];
                }
                update[i] = current;
            }

            current = current.forward[0];

            if (current != null && current.key == key) {
                for (int i = 0; i <= level; i++) {
                    if (update[i].forward[i] != current) {
                        break;
                    }
                    update[i].forward[i] = current.forward[i];
                }

                while (level > 0 && header.forward[level] == null) {
                    level--;
                }
            }
        }

        /**
         * Display SkipList structure.
         */
        public void display() {
            System.out.println("SkipList structure:");
            for (int level = this.level; level >= 0; level--) {
                SkipListNode current = header.forward[level];
                String levelStr = "Level " + level + ": ";
                while (current != null) {
                    levelStr += current.key + " -> ";
                    current = current.forward[level];
                }
                levelStr += "None";
                System.out.println(levelStr);
            }
        }
    }

    public static void main(String[] args) {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();
        long memoryBefore = runtime.totalMemory() - runtime.freeMemory();
        long startTime = System.nanoTime();
        
        // Create SkipList instance
        SkipListImpl skipList = new SkipListImpl();
        
        // Insert elements
        int[] elements = {3, 6, 7, 9, 12, 19, 17, 26, 21, 25};
        for (int elem : elements) {
            skipList.insert(elem);
        }

        // Display structure
        skipList.display();
        
        // Search operations
        int[] searchKeys = {19, 15, 21};
        System.out.println("\nSearch results:");
        for (int key : searchKeys) {
            Integer result = skipList.search(key);
            System.out.println("Search for " + key + ": " + (result != null ? "Found" : "Not found"));
        }
        
        // Delete operation
        skipList.delete(17);
        System.out.println("\nAfter deleting 17:");
        skipList.display();
        
        long endTime = System.nanoTime();
        long memoryAfter = runtime.totalMemory() - runtime.freeMemory();
        long memoryUsed = memoryAfter - memoryBefore;
        
        // Performance statistics
        System.out.println("\n--- Performance Statistics ---");
        System.out.printf("Execution time: %.4f ms%n", (endTime - startTime) / 1_000_000.0);
        System.out.println("Memory usage: " + (memoryUsed > 0 ? memoryUsed : "negligible") + " bytes");
        System.out.println("\nTime Complexity (Average):");
        System.out.println("  - Search: O(log n)");
        System.out.println("  - Insert: O(log n)");
        System.out.println("  - Delete: O(log n)");
        System.out.println("  - Space: O(n)");
    }
}